File: 2218-maximum-value-of-k-coins-from-piles/2218-maximum-value-of-k-coins-from-piles.js
/**
 * @param {number[][]} piles
 * @param {number} k
 * @return {number}
 */

var maxValueOfCoins = function(piles, k) {
    // 实际为凑硬币问题或者部分背包问题的变种
    //从优化角度考虑为如何维护最少信息，求得后续。实际上对于此类题完全不用知道到底怎么选的。
    //只要考虑一个pile选完之后的影响即可
    
    var n = piles.length
    var dp = Array(n).fill().map(() => Array(k + 1).fill(0));
    var curSum = 0
    var ans = 0
    var i = 1
    for (var elem of piles[0]) {
        curSum += elem
        dp[0][i] = curSum
        ans = Math.max(ans, dp[0][k])
        i++
    }

    for (var i = 1; i < n; i++) {
        var pile = piles[i]
        var curSum = 0
        for (var j = 0; j < pile.length + 1; j++){
            var maxK = k - j
            if (j > 0) {
                curSum += pile[j - 1]
            }
            for (var prev = 0; prev <= maxK; prev++) {
                dp[i][j + prev] = Math.max(dp[i][j + prev], curSum + dp[i - 1][prev])
        }
        }
        ans = Math.max(ans, dp[i][k])
    }

    return ans
    
    
};

File: 2218-maximum-value-of-k-coins-from-piles/2218-maximum-value-of-k-coins-from-piles.py
class Solution:
    def maxValueOfCoins(self, piles: List[List[int]], k: int) -> int:
        self.memo = defaultdict(int)
        self.ans = 0
        self.k = k
        self.find(len(piles) - 1, k, piles)

        return self.ans
    
    def find(self, i, remainingK,  piles):
        key = tuple([i, remainingK])

        if key in self.memo:
            return self.memo[key]
        
        if i == -1:
            return 0
        if remainingK < 0:
            return 0
        pile = piles[i]
        n = len(pile)
        
        if i == 0 and n < remainingK:
            return 0
        if i == 0:
            self.ans = max(self.ans, sum(pile[:remainingK]))
            self.memo[key] = sum(pile[:remainingK])
            return self.memo[key]
        
        maxSelected = min(n, remainingK)
        curRes = 0
        curSum = 0
        
        for selectedNum in range(maxSelected + 1):
            curVal = 0
            if selectedNum > 0:
                curVal = pile[selectedNum - 1]
            curSum += curVal
            curRes = max(self.find(i - 1, remainingK - selectedNum, piles) + curSum, curRes)

        self.memo[key] = curRes
        
        if remainingK == self.k:
            self.ans = max(self.ans, curRes)
        return curRes
                
            
        
            
    
            
            
        
        
        